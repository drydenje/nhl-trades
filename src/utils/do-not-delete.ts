// might need this to remove "no inputs were found in config file"
// error in tsconfig.json
// doesn't do anything else
const a = 1;
