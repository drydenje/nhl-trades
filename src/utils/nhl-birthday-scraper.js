// Select the first player without a birthdate
// Fetch the birthday from nhle.com
// Return playerArray with the new data

const addBirthdayToPlayer = async (playerArray) => {
  const playerToScrape = playerArray.filter(
    (player) => player.birthDate === undefined
  );

  const NHL_PLAYER_DETAILS_ENDPOINT = `https://api-web.nhle.com/v1/player/${playerToScrape[0].id}/landing`;

  const playerDetails = await fetch(NHL_PLAYER_DETAILS_ENDPOINT).then(
    (response) => response.json()
  );

  const newArray = playerArray.map((player) => {
    let bd = undefined;
    if (playerToScrape[0].id === player.id)
      bd = { birthDate: playerDetails.birthDate };

    return {
      ...player,
      ...bd,
    };
  });

  return newArray;
};

export { addBirthdayToPlayer };
