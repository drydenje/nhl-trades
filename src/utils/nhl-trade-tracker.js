const cheerio = require("cheerio");

// Given a full html page, it will count the number of
// pages to scrape for that year
const getPageCount = async (page) => {
  let $ = cheerio.load(page);
  const pageCount = $(".pagination").find("a:last").prev().text();

  return parseInt(pageCount) || 1;
};

const getTradesFromPage = async (page) => {
  const $ = cheerio.load(page);
  const lines = [];
  $("#container")
    .find("> table")
    .each((index, trade) => {
      lines.push(parseTrade(trade));
    });

  return lines;
};

// Pull all relevent data from the specific trade, and return an object
// {
//     "date": "April 12, 2021",
//     "teams": {
//         "Washington Capitals": [
//             {
//                 "name": "Anthony Mantha",
//                 "hockeyDBid": "136790"
//             }
//         ],
//         "Detroit Red Wings": [
//             {
//                 "name": "2021 1st round pick"
//             },
//             {
//                 "name": "Richard Panik",
//                 "hockeyDBid": "108683"
//             },
//             {
//                 "name": "Jakub Vrana",
//                 "hockeyDBid": "161382"
//             },
//             {
//                 "name": "2022 2nd round pick"
//             }
//         ]
//     },
//     "comment": null
// }
const parseTrade = (html) => {
  let $ = cheerio.load(html);
  const headings = $("tbody > tr")
    .find("td > strong")
    .map(function () {
      if ($(this).text() !== "Date") {
        return $(this).text().replace(" acquire", "");
      }
    })
    .toArray();

  const date = $("tr:nth(1) > td:nth(1)").text().trim();

  const roster = [];
  for (let i = 0; i <= 1; i++) {
    roster[i] = $(`table:nth(${i + 1})`)
      .text()
      .trim()
      .split("\n")
      .map((item) => {
        const playerName = item.trim();
        const r = /([0-9])\w+/g;
        const playerID = $(`a:contains(${playerName})`).attr("href");
        const player = {
          name: playerName,
        };
        playerID ? (player.hockeyDBid = playerID.match(r)[0]) : null;
        return player;
      })
      .filter((s) => s.name);
  }

  const c = $(`.comment`).next().text();
  const comment = c ? c : null;
  const teams = {};
  teams[headings[0]] = roster[0];
  teams[headings[1]] = roster[1];

  return {
    date,
    teams,
    comment,
  };
};

// this will grab  a list of all the nhl trade seasons
const getSeasonList = (page) => {
  let $ = cheerio.load(page);
  const arr = $(".s_list > tbody")
    .find("tr")
    .text()
    .split("\n")
    .filter((s) => s)
    .map((s) => s.trim());

  const seasons = arr.reduce((accumulator, value) => {
    return { ...accumulator, [value]: [] };
  }, {});

  return {
    seasons,
  };
};

export { getPageCount, getSeasonList, getTradesFromPage, parseTrade };
