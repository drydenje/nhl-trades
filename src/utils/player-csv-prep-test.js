import { addHrIdsToPlayerList } from "./player-csv-prep-test.js";
import { server, http, HttpResponse } from "../config/testServer";

const DATA_FILEPATH = `./public/scraped-data/`;

// describe("add-nhl-players function", () => {
//   // this shouldn't return csv?
//   test("should return a csv (file?) given an array of nhl.com players", () => {
//     const headings = ["id", "name", "hrId"];
//     const players = [
//       {
//         id: 8444924,
//         name: "Jim Anderson",
//       },
//       {
//         id: 8444926,
//         name: "Murray Anderson",
//       },
//       {
//         id: 8444995,
//         name: "Pete Backor",
//       },
//     ];

//     // const yearToCheck = "2021-22";
//     // const result = checkArray(seasons[yearToCheck]);
//     // expect(result).toEqual(1);
//   });
// });

describe("add-hr-players function", () => {
  test("should add hockey-reference.com id's to the player array", () => {
    const players = [
      {
        id: 8444924,
        name: "Jim Anderson",
      },
      {
        id: 8444926,
        name: "Murray Anderson",
      },
      {
        id: 8444995,
        name: "Pete Backor",
      },
    ];

    const hrPlayers = [
      { name: "Jim Anderson", hrID: "anderji01" },
      { name: "Murray Anderson", hrID: "andermu01" },
      { name: "Pete Backor", hrID: "backope01" },
      { name: "Mats Sundin", hrID: "sundima01" },
      { name: "St√©phane Fiset", hrID: "fisetst01" },
    ];

    // const yearToCheck = "2021-22";
    const result = checkArray(seasons[yearToCheck]);
    // expect(result).toEqual(1);
  });
});
