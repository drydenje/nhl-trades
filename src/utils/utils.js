const fs = require("fs");
import cliProgress from "cli-progress";
import cheerio from "cheerio";
import latinize from "latinize";

// checks to see if an array exists and is empty.
// Returns 0 if empty, 1 if the array has data, and 'false' if it's not an array
const checkArray = (arr) => {
  return Array.isArray(arr) && arr.length;
};

// takes nhlPlayer and another Player object
const checkNameAndCity = (
  { firstName, lastName, birthCity },
  { name: hName, birthCity: hBirthCity }
) => {
  return (
    `${firstName.default} ${lastName.default}` === hName &&
    birthCity.default === hBirthCity
  );
};

// takes nhlPlayer and another Player object
const checkNameAndBirthDate = (
  { firstName, lastName, birthDate },
  { name: hName, birthDate: hBirthDate }
) => {
  return (
    `${firstName.default} ${lastName.default}` === hName &&
    birthDate === hBirthDate
  );
};

// Returns the next year to scrape
const getNextYear = (obj) => {
  const yearToScrape = Object.entries(obj).filter((year) => {
    if (checkArray(year[1]) === 0) {
      return year;
    }
  });

  if (yearToScrape.length > 0) {
    // return the year of the first object of the array
    return yearToScrape[0][0];
  } else {
    return false;
  }
};

// Returns a single html page from the given url
const getPage = async (url) => {
  try {
    const res = await fetch(url);
    if (!res.ok) {
      throw new Error(`Request failed with status code ${result.status}`);
    }
    const data = await res.text();
    return data;
  } catch (error) {
    console.error(`Could not fetch page: ${error}`);
  }
};

// reads the file from disk
const readFile = (filePath) => {
  try {
    const data = fs.readFileSync(filePath, "utf8");

    return data;
  } catch (error) {
    console.error(error);
  }
};

// writes the file to disk
const writeFile = (filePath, data) => {
  try {
    fs.writeFileSync(filePath, JSON.stringify(data));
  } catch (error) {
    console.error(error);
  }
};

const removeNickname = (nickname) => {
  // 1. Remove the nickname and the quotes around it
  // 2. This leaves two whitespace characters
  // 3. Replace the two characters with a single one
  return nickname.replace(/"([^"]*)"/g, "").replace(/\s\s+/g, " ");
};

const combineSiteIds = (nhlPlayers, hdbPlayers, hrPlayers) => {
  const players = JSON.parse(readFile(nhlPlayers));
  const hdbIDs = JSON.parse(readFile(hdbPlayers));
  const hrIDs = JSON.parse(readFile(hrPlayers));

  const progressBar = new cliProgress.SingleBar(
    {},
    cliProgress.Presets.shades_classic
  );
  progressBar.start(players.length, 0);

  //512

  const playerResult = players.map((player, index) => {
    const fullName = `${player.firstName.default} ${player.lastName.default}`;
    // const singleNameCheck = players.filter()
    const hr = hrIDs.find(
      (hrPlayer) =>
        checkNameAndBirthDate(player, hrPlayer) ||
        checkNameAndCity(player, hrPlayer)
    );

    const hdb = hdbIDs.find(
      (hdbPlayer) =>
        checkNameAndBirthDate(player, hdbPlayer) ||
        checkNameAndCity(player, hdbPlayer)
    );

    progressBar.update(index + 1);

    return {
      ...player,
      hrID: hr ? hr.id : null,
      hdbID: hdb ? hdb.hdbID : null,
    };
  });

  progressBar.stop();

  // return an array of objects containing all players with hdbid and hrids added
  return playerResult;
};

export {
  checkArray,
  combineSiteIds,
  getNextYear,
  getPage,
  readFile,
  writeFile,
  removeNickname,
};
