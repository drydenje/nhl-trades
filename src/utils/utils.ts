const cheerio = require("cheerio");
const axios = require("axios");

const getTradeTeams = async (url: string) => {
  const axiosResponse = await axios.request({
    method: "GET",
    url: url,
    headers: {
      "User-Agent":
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36",
    },
  });

  const $ = cheerio.load(axiosResponse.data);
  const lines = [];
  const tables = $("#container")
    // .find("> table")
    .find("> table")
    // .children("tbody")
    .each((index, trade) => {
      //console.log(`ELEMENT: #${index}`);
      // console.log($(trade).find("tbody > tr > td > strong").html());
      // console.log($(trade).html());
      // lines.push(extract(trade));
      // console.log(extract(trade))
      // return extract(trade);
      lines.push(parseTrade(trade));
    });
  // .text()
  // .trim();
  // .toArray();
  // console.log(lines);
  return lines;
};

const parseTrade = (input: string) => {
  let $ = cheerio.load(input);
  const headings = $("tbody > tr")
    .find("td > strong")
    .map(function () {
      if ($(this).text() !== "Date") {
        return $(this).text().replace(" acquire", "");
      }
    })
    .toArray();

  const date = $("tr:nth(1) > td:nth(1)").text().trim();

  const roster = [];
  for (let i = 0; i <= 1; i++) {
    roster[i] = $(`table:nth(${i + 1})`)
      .text()
      .trim()
      .split("\n")
      .map((item) => {
        return item.trim();
      })
      .filter((s) => s);
  }

  const teams = {};
  teams[headings[0]] = roster[0];
  teams[headings[1]] = roster[1];

  return {
    date,
    teams,
  };
};

export { getTradeTeams };
