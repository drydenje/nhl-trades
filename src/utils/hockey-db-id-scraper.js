import { readFile, removeNickname } from "./utils";
const path = require("path");
const cheerio = require("cheerio");
const chalk = require("chalk");

// can probably delete this
const parsePlayer = (html) => {
  const $ = cheerio.load(html);
  // console.log(html);
  const playerName = $("a").text().trim();
  const playerId = $("a").attr("href").match(/\d+$/)[0];
  const birthDate = $("td");
  // console.log("bd:", birthDate);

  const data = {
    name: playerName,
    id: playerId,
    birthDate: null,
  };

  return data;
};

const getNamesFromPage = (page) => {
  const $ = cheerio.load(page);
  const names = $(".sortable > tbody")
    .children()
    .map((index, player) => {
      return {
        hdbID: $("a", player).attr("href").match(/\d+$/)[0],
        name: removeNickname($("a", player).html()),
        birthDate: new Date(
          $("td:eq(2)", player).html().trim()
        ).toLocaleDateString("en-CA"),
        // new Date('02/04/1930').toLocaleDateString('en-CA')
      };
    })
    .toArray();
  // console.log("N:", names);
  return names;
};

const play = [
  [
    {
      birthDate: "12/26/1937",
      id: "5512",
      name: "Gene Ubriaco",
    },
    {
      birthDate: "12/31/1992",
      id: "123037",
      name: "Dominik Uher",
    },
    {
      birthDate: "10/01/1969",
      id: "5515",
      name: "Igor Ulanov",
    },
  ],
  [
    {
      birthDate: "12/26/1935",
      id: "5516",
      name: "Norm Ullman",
    },
    {
      birthDate: "07/31/1993",
      id: "152794",
      name: "Linus Ullmark",
    },
    {
      birthDate: "04/22/1989",
      id: "112734",
      name: "David Ullstrom",
    },
  ],
  [
    {
      birthDate: "04/27/1977",
      id: "27868",
      name: "Jeff Ulmer",
    },
    {
      birthDate: "09/14/1980",
      id: "31019",
      name: "Layne Ulmer",
    },
    {
      birthDate: "05/03/1982",
      id: "54470",
      name: "R.J. Umberger",
    },
  ],
];

const getAllPlayers = (letters, baseFilePath) => {
  let allPlayers = [];
  // for each letter in the array
  letters.forEach((letter) => {
    const fileToParse = `${baseFilePath}/NHL Player List -- ${letter.toUpperCase()}.html`;
    const html = readFile(fileToParse);
    const playersForLetter = getNamesFromPage(html);
    allPlayers = [...allPlayers, ...playersForLetter];
  });
  return allPlayers.flat(); //.flatMap(player => );
};

// Takes two players and returns one combined (if it's the same person)
const setHdbID = (player, hdbPlayer) => {
  // check that the name and birthdate match to verify it's the same person
  if (
    player.name === hdbPlayer.name &&
    player.birthDate === hdbPlayer.birthDate
  ) {
    // it's the same person, add the hdbID
    return {
      ...player,
      hdbID: hdbPlayer.hdbID ? hdbPlayer.hdbID : null,
    };
  } else {
    // it's a different person, just return the first player without the hdbID
    return {
      ...player,
    };
  }
};

const addHdbIDToAll = (players, hdbPlayers) => {
  const result = players.map((player) => {
    const findHdbPlayers = hdbPlayers.filter(
      (hdbPlayer) => hdbPlayer.name === player.name
    );
    // console.log("FIND:", findHdbPlayers);
    // console.log("FIND:", player);

    // if (findHdbPlayers.length === 0) {
    //   console.log("ZERO:", findHdbPlayers);
    //   console.log("PL:", player);
    // }
    // return player;

    const fPlayers = findHdbPlayers.filter(
      (p) => p.birthDate === player.birthDate
    );
    // console.log("FP:", fPlayers);

    return fPlayers[0] || player;
    // return setHdbID(player, temp);
  });

  return result;
};

export {
  getNamesFromPage,
  parsePlayer,
  getAllPlayers,
  setHdbID,
  addHdbIDToAll,
};
