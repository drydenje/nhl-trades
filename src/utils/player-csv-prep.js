import { convertArrayToCSV } from "convert-array-to-csv";
import { writeFile } from "./utils";

const addHrIdsToPlayerList = (players, hrPlayers) => {
  const duplicates = [];

  // the list of players that will be returned
  let finalPlayersArray = players.map((player) => {
    console.log("Player:", player);
    // for each player in the nhl.com list

    // select the player from the hr list
    const hrFindPlayer = hrPlayers.filter((hrPlayer) => {
      const comparison = hrPlayer.name.localeCompare(player.name, "en-US", {
        sensitivity: "base",
      });

      return comparison ? false : true;
    });

    // console.log("FIND:", hrFindPlayer);
    if (hrFindPlayer.length > 1) {
      duplicates.push(player);
      return {
        ...player,
      };
    } else {
      return {
        ...player,
        ...(hrFindPlayer[0] ? { hrID: hrFindPlayer[0].hrID } : null),
      };
    }
  });

  writeFile(`./public/scraped-data/duplicate-ids.json`, duplicates);
  console.log("Duplicates:", duplicates);

  return finalPlayersArray;
};

export { addHrIdsToPlayerList };
