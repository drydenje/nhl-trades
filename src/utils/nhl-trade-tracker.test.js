import { getPageCount } from "./nhl-trade-tracker";
import { readFile } from "./utils";
// import page from "../../public/raw-mock-data/1977-78-1.html";

const page = readFile("./public/raw-mock-data/1977-78-1.html");
// console.log(page);

describe("getPageCount function", () => {
  test.only("should return the number of pages to scrape for that year, given an html page", async () => {
    const result = await getPageCount(page);
    console.log("R:", result);
    expect(result).toEqual(3);
  });
});
