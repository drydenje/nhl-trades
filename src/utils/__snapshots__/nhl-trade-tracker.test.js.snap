// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAllTradesForYear function should return an html page when passed a url 1`] = `false`;
